File: web-sdk/apps/lex-looter/src/game/notation.ts
export const BOARD_COLS = 40;
export const BOARD_ROWS = 25;
export const CELL_SIZE = 20;
export const CANVAS_WIDTH = BOARD_COLS * CELL_SIZE;
export const CANVAS_HEIGHT = BOARD_ROWS * CELL_SIZE;

export type BoardCell = {
	col: number;
	row: number;
};

export type PixelPoint = {
	x: number;
	y: number;
};

export const CORNER_NOTATION = {
	tl: 'A25',
	tr: 'AN25',
	bl: 'A1',
	br: 'AN1',
} as const;

export const colToLabel = (col: number) => {
	if (col < 0) throw new Error(`Column out of range: ${col}`);
	let label = '';
	let value = col;
	while (true) {
		label = String.fromCharCode('A'.charCodeAt(0) + (value % 26)) + label;
		value = Math.floor(value / 26) - 1;
		if (value < 0) return label;
	}
};

export const labelToCol = (label: string) => {
	if (!/^[a-z]+$/i.test(label)) throw new Error(`Invalid column label: ${label}`);
	return [...label.toUpperCase()].reduce(
		(total, char) => total * 26 + (char.charCodeAt(0) - 'A'.charCodeAt(0) + 1),
		0,
	) - 1;
};

export const cellToNotation = ({ col, row }: BoardCell) => {
	if (col < 0 || col >= BOARD_COLS || row < 0 || row >= BOARD_ROWS) {
		throw new Error(`Cell out of range: ${col}, ${row}`);
	}
	return `${colToLabel(col)}${row + 1}`;
};

export const notationToCell = (notation: string): BoardCell => {
	const match = notation.match(/^([a-z]+)(\d+)$/i);
	if (!match) throw new Error(`Invalid notation: ${notation}`);
	const col = labelToCol(match[1]);
	const row = Number(match[2]) - 1;
	if (col < 0 || col >= BOARD_COLS || row < 0 || row >= BOARD_ROWS) {
		throw new Error(`Notation out of range: ${notation}`);
	}
	return { col, row };
};

export const cellToPixelCenter = ({ col, row }: BoardCell): PixelPoint => ({
	x: (col + 0.5) * CELL_SIZE,
	y: CANVAS_HEIGHT - (row + 0.5) * CELL_SIZE,
});

export const notationToPixelCenter = (notation: string): PixelPoint =>
	cellToPixelCenter(notationToCell(notation));

export const pixelToNotation = ({ x, y }: PixelPoint) =>
	cellToNotation({
		col: Math.min(Math.max(Math.floor(x / CELL_SIZE), 0), BOARD_COLS - 1),
		row: Math.min(Math.max(Math.floor((CANVAS_HEIGHT - y) / CELL_SIZE), 0), BOARD_ROWS - 1),
	});

File: math-sdk/games/lexlooter/notation.py
"""Board notation helpers for Lex Looter.

Rows are bottom-origin like chess: A1 is bottom-left, AN25 is top-right.
Canvas pixels are top-origin, so y conversion is inverted.
"""

BOARD_COLS = 40
BOARD_ROWS = 25
CELL_SIZE = 20
CANVAS_WIDTH = BOARD_COLS * CELL_SIZE
CANVAS_HEIGHT = BOARD_ROWS * CELL_SIZE

CORNER_NOTATION = {
    "tl": "A25",
    "tr": "AN25",
    "bl": "A1",
    "br": "AN1",
}


def col_to_label(col: int) -> str:
    if col < 0:
        raise ValueError(f"Column out of range: {col}")

    label = ""
    value = col
    while True:
        label = chr(ord("A") + (value % 26)) + label
        value = value // 26 - 1
        if value < 0:
            return label


def label_to_col(label: str) -> int:
    if not label or not label.isalpha():
        raise ValueError(f"Invalid column label: {label}")

    value = 0
    for char in label.upper():
        value = value * 26 + (ord(char) - ord("A") + 1)
    return value - 1


def cell_to_notation(col: int, row: int) -> str:
    if col < 0 or col >= BOARD_COLS or row < 0 or row >= BOARD_ROWS:
        raise ValueError(f"Cell out of range: col={col}, row={row}")
    return f"{col_to_label(col)}{row + 1}"


def notation_to_cell(notation: str) -> dict:
    letters = "".join(char for char in notation if char.isalpha())
    numbers = "".join(char for char in notation if char.isdigit())
    if not letters or not numbers:
        raise ValueError(f"Invalid notation: {notation}")

    col = label_to_col(letters)
    row = int(numbers) - 1
    if col < 0 or col >= BOARD_COLS or row < 0 or row >= BOARD_ROWS:
        raise ValueError(f"Notation out of range: {notation}")
    return {"col": col, "row": row}


def cell_to_pixel_center(col: int, row: int) -> dict:
    return {
        "x": (col + 0.5) * CELL_SIZE,
        "y": CANVAS_HEIGHT - (row + 0.5) * CELL_SIZE,
    }


def pixel_to_notation(x: float, y: float) -> str:
    col = min(max(int(x // CELL_SIZE), 0), BOARD_COLS - 1)
    row = min(max(int((CANVAS_HEIGHT - y) // CELL_SIZE), 0), BOARD_ROWS - 1)
    return cell_to_notation(col, row)


def notation_to_normalized(notation: str) -> tuple[float, float]:
    cell = notation_to_cell(notation)
    pixel = cell_to_pixel_center(cell["col"], cell["row"])
    return (
        round(pixel["x"] / CANVAS_WIDTH, 4),
        round(pixel["y"] / CANVAS_HEIGHT, 4),
    )


def board_payload() -> dict:
    return {
        "cols": BOARD_COLS,
        "rows": BOARD_ROWS,
        "cellSize": CELL_SIZE,
        "width": CANVAS_WIDTH,
        "height": CANVAS_HEIGHT,
        "corners": CORNER_NOTATION,
    }
